addon/-private/class/modifier.ts

```diff
@@ -60,27 +66,87 @@ export const _implementsLegacyHooks = <S>(
  * update if any of those values change.
  */
 export default class ClassBasedModifier<S = DefaultSignature> {
+  // Done this way with the weird combination of `declare` and `defineProperty`
+  // so that subclasses which are overriding this by writing their own `args`
+  // field type declarations continue to type check correctly. (If we introduced
+  // a getter here, existing classes defining their args via a `declare args:`
+  // would stop type checking, because TS -- correctly! -- differentiates
+  // between class fields and getters).
   /**
    * The arguments passed to the modifier. `args.positional` is an array of
    * positional arguments, and `args.named` is an object containing the named
    * arguments.
+   *
+   * @deprecated Until 4.0. Access positional and named arguments directly in
+   *   the `modify` hook instead.
    */
-  readonly args: ArgsFor<S>;
+  declare readonly args: ArgsFor<S>;
 
+  // Done this way with the weird combination of `declare` and `defineProperty`
+  // so that subclasses which are overriding this by writing their own `element`
+  // field declarations continue to type check correctly.
   /**
    * The element the modifier is applied to.
    *
    * @warning `element` is ***not*** available during `constructor` or
    *   `willDestroy`.
+   * @deprecated Until 4.0. Access the `element` as an argument in the `modify`
+   *   hook instead.
    */
-  // SAFETY: this is managed correctly by the class-based modifier. It is not
-  // available during the `constructor`.
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  element: ElementFor<S> = null as any;
+  declare element: ElementFor<S>;
 
   constructor(owner: unknown, args: ArgsFor<S>) {
     setOwner(this, owner);
-    this.args = args;
+
+    // SAFETY: the point here is (for the period where we are providing `args`
+    // and `element`) to provide an internal-only way of setting and update the
+    // `args` for the modifier instance; we use the `InternalClassBasedModifier`
+    // interface to represent the internal-only API in a way that end users do
+    // *not* have access to when subclassing `ClassBasedModifier`.
+    (this as unknown as InternalClassBasedModifier<S>)[Args] = args;
+
+    // TODO: remove both of these `defineProperty`s at v4.
+    Object.defineProperty(this, 'args', {
```

Comment:
I'm curious, why not make it a getter in the class body (specifically, why use `Object.defineProperty` for **every** instance)?

I would expect this to work the same:

```js
class Modifier {
  get args() {
    deprecate(...);
    return this[Args];
  }
}
```

Reply:
The comment on the field itself explains: we have to do this because existing subclasses used the fields as a place to define the types for the subclasses, and `field: Type` is a different type than `get field: Type`.

Reply:
That said, we *should* do this as `Object.defineProperty(Modifier.prototype, 'args', ...)` instead so we don't pay the perf hit!